File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    answer = 0
    n = len(words)
    visited = [False] * (n+1)
    
    def strcmp(a, b):
        cnt = 0
        for i in range(len(a)):
            if a[i] == b[i]:
                cnt += 1
        if cnt == len(a)-1:
            return True
        else: return False
    
    que = deque()
    que.append((begin, 0))
    visited[0] = True

    while que:
        cur_v, cur_len = que.popleft()
        if cur_v == target:
            answer = cur_len
            break

        for i in range(len(words)):
            if strcmp(cur_v, words[i]) and not visited[i+1]:
                que.append((words[i], cur_len+1))
                visited[i+1] = True        
    
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;

        answer = bfs(begin, target, words);

        if (answer == -1) {
            return 0;
        }

        return answer;
    }

    public int bfs(String begin, String target, String[] words) {
        Queue<String> queue = new LinkedList<>();
        queue.add(begin);

        Set<String> visited = new HashSet<>();
        visited.add(begin);

        int level = 0;
        while (!queue.isEmpty()) {
            int len = queue.size();
            for (int i = 0; i < len; i++) {
                String cur = queue.poll();
                if (cur.equals(target)) {
                    return level;
                }

                for (String word : words) {
                    if (check(cur, word) && !visited.contains(word)) {
                        queue.add(word);
                        visited.add(word);
                    }
                }
            }
            level++;
        }

        return -1;
    }

    public boolean check(String cur, String word) {
        int cnt = 0;
        for (int i = 0; i < cur.length(); i++) {
            if (cur.charAt(i) != word.charAt(i)) {
                cnt++;
            }
        }

        if (cnt == 1) {
            return true;
        }

        return false;
    }
}